[No QA] Created IOU preview component within chat

tugbadogan opened the pull request:
@iwiznia @Julesssss 

### Details

- Added IOU Preview component
- Fetched required iou data from Onyx
- Implemented the Quote section 
- Implemented IOU Preview UI

### Fixed Issues
Fixes https://github.com/Expensify/Expensify.cash/issues/2027

### Tests

- Logged in to account Tugba Oge and requested money from user Fadime Dogan.
- Verified quote and preview box working as expected.
- Paid money via API call.
- Once IOU is settled, verified that there is a separate quoted message about settlement and there is no preview box.

### QA Steps

For now, this is only testable internally. The cancel/decline cases will be tested as part of the IOU Details PR, which will introduce that functionality. For now, I have been testing this with API calls. 

- Login to user X
- Create IOU transaction with user Y via UI or API call as mentioned in issue description https://github.com/Expensify/Expensify.cash/issues/2027
- Open 1:1 chat with the person from transaction
- Observe that there is a quoted message about the request and preview box with `amount` and `Y owes X`  message.
- Login to user Y
- Open 1:1 chat with user X
-  Observe that there is a quoted message about the request and preview box with `amount`, `Y owes X`  message and `Pay` button.
- Pay the IOU transaction with API call
- For both users, observe that there is a quoted message about the settlement. Make sure there is no preview box.
- Create a new transaction
- Cancel the transaction
- For both users, the Preview component should not display

### Tested On

- [X] Web
- [X] Mobile Web
- [X] Desktop
- [X] iOS
- [X] Android

### Screenshots

#### Web

Owner / Manager Views
<img width="400" src="https://user-images.githubusercontent.com/3853003/114093137-aa147480-98b2-11eb-83be-968ce1b0ae88.png"> <img width="400" src="https://user-images.githubusercontent.com/3853003/114093148-ad0f6500-98b2-11eb-91d8-da1765efdc45.png">

Group Chat View
<img width="400" alt="Screenshot 2021-03-30 at 16 16 21" src="https://user-images.githubusercontent.com/3853003/113062959-764b9780-91ac-11eb-98fc-bce9a6866415.png">

#### Mobile Web

Owner / Manager Views
<img width="352" src="https://user-images.githubusercontent.com/3853003/114093381-f364c400-98b2-11eb-9eaa-82bba9f2e4db.png"> <img width="346" src="https://user-images.githubusercontent.com/3853003/114093375-f2339700-98b2-11eb-8223-4f09c4381592.png">

#### Desktop

Owner / Manager Views
<img width="400" src="https://user-images.githubusercontent.com/3853003/114093557-2b6c0700-98b3-11eb-89e8-471a241348b6.png"> <img width="400" src="https://user-images.githubusercontent.com/3853003/114093564-2dce6100-98b3-11eb-8496-6984244456bb.png">

#### iOS

Owner / Manager Views
<img width="400" src="https://user-images.githubusercontent.com/3853003/114093685-53f40100-98b3-11eb-8b3b-d4c1371fbe82.png"> <img width="400" src="https://user-images.githubusercontent.com/3853003/114093693-55bdc480-98b3-11eb-8151-a7c3c3ffef36.png">

#### Android

Owner / Manager Views
<img width="400" src="https://user-images.githubusercontent.com/3853003/114093840-83a30900-98b3-11eb-8810-62f5db8de843.png"> <img width="400" src="https://user-images.githubusercontent.com/3853003/114093836-81d94580-98b3-11eb-9e3c-a3b9bc5c8196.png">

tugbadogan committed:
Created IOU preview component within chat

tugbadogan committed:
Created IOU preview component within chat

tugbadogan committed:
Added ReportActionItemIOUQuote component

tugbadogan committed:
Added ReportActionItemIOUQuote component

Julesssss commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -27,9 +27,19 @@ const propTypes = {
     betas: PropTypes.arrayOf(PropTypes.string),
+
+    // The report currently being looked at
+    report: PropTypes.shape({
+
+        // IOU report ID associated with current report
+        iouReportID: PropTypes.number,
```
Any reason not to just pass the reportID at this stage, and subscribe to the full report Onyx further down the Component chain?

tugbadogan replied:
Sure, I can pass just iouReportID here. 👍

marcaaron replied:
I'm not sure this was addressed and I think @Julesssss was suggesting that we pass the `iouReportID` directly instead of subscribing to the entire report object.

Julesssss commented on src/components/ReportActionItemIOUQuote.js:
```diff
@@ -0,0 +1,30 @@
+            const viewDetails = '<br /><a href="#" style="text-decoration:none;">View Details</a>';
+            const html = `<blockquote>${fragment.text}${viewDetails}</blockquote>`;
+            return (
+                <RenderHTML key={`iouQuote-${action.sequenceNumber}-${index}`} html={html} debug={false} />
```
is the debug prop not false by default?

Julesssss reviewed (commented):
Looking good so far. 

Just to let you know, until now group chat IOU actions have been of type `ADDCOMMENT`, but we've changed this to be an `IOU` action, which should make it easier to render the Quote component in group chats.

[This change is currently on staging.expensify.com]

tugbadogan commented:
> Just to let you know, until now group chat IOU actions have been of type `ADDCOMMENT`, but we've changed this to be an `IOU` action, which should make it easier to render the Quote component in group chats.

Do we use the same API method for creating transactions in group chat? Should I worry about different transaction types, request money and split bill?

Julesssss commented:
> Do we use the same API method for creating transactions in group chat? Should I worry about different transaction types, request money and split bill?

Good question, sorry I didn't share this to begin with: 
- Creating a group split is done with [this API request](https://github.com/Expensify/Expensify.cash/issues/1988#issuecomment-808296948) (you will first need to create a report for these users [via this API](https://github.com/Expensify/Expensify.cash/blob/1902682e6a63843853f8919930cc2bad893a5cc9/src/libs/API.js#L264-L269) and pass that reportID)
- This simply creates creates a 1:1 IOU for each individual participant and the bill creator (so in the example above, 2 IOU reports are created: `jules | jules+4`, `jules | jules+6`)
- No need to worry about different types, once the above request is made, you will still interact with IOUActions in a 1:1 report in the usual way
- The exception is that within a group chat the IOUAction contains participant names -- which you can use to display all participants in the quote message. The Preview Component within group chats should look like the following (no View Details link, all participants listed):  

![Chat (9)](https://user-images.githubusercontent.com/10736861/112835825-b4a06400-9091-11eb-937d-948127426f89.png)
![FAB Open (2)](https://user-images.githubusercontent.com/10736861/112835828-b538fa80-9091-11eb-997e-c5d9f13b291a.png)

tugbadogan commented:
Hey @Julesssss 

When I called `PayIOU` API method, `iouReportId` key is removed from report, `hasOutstandingIOU` key is set to `true` and  `reportIOUs_` object is deleted from Onyx. So, I cannot fetch state, status and I cannot display the paid amount on the preview box as I don't have access to the transactions anymore. 

Is there anywhere else that I can find settled IOUs, so I can display the paid amount?

Julesssss commented:
Actually ignore that last part, the view details link isn't part of this PR 🙃

tugbadogan committed:
Added IOU Preview UI

tugbadogan committed:
Added IOU Preview UI

tugbadogan commented:
I updated the PR. It's ready for review.

Julesssss commented:
Nice work! Here are some things I've noticed: 

1. Once an IOU is settled we don't need to render the component
I think we need to check `isMostRecentIOUReport` AND `report.hasOutstandingIOU`

![Screenshot 2021-03-31 at 14 43 37](https://user-images.githubusercontent.com/10736861/113156118-996c5a80-9231-11eb-9523-aa303bdc8aab.png)

2. If you don't have a reason, we should remove the 'for': (Requested $9 from Jules **for**)

![Screenshot 2021-03-31 at 15 07 17](https://user-images.githubusercontent.com/10736861/113157492-da18a380-9232-11eb-997a-c49f50022da9.png)

3. When a transaction is cancelled, we don't seem to be immediately updating other IOUs. 
    - the middle component should be hidden, (this happens after a refresh)
    - ![device-2021-03-31-151535](https://user-images.githubusercontent.com/10736861/113158981-29130880-9234-11eb-8244-158f53c2fbb9.png)
    - the preview component should say $5, not $10 -- I can see the `cachedTotal` is correctly showing $5
    - ![Screenshot 2021-03-31 at 15 21 50](https://user-images.githubusercontent.com/10736861/113159894-dede5700-9234-11eb-8320-f99aef543f23.png)

_Also, we should just show a users first name for the requested quote message (but this is a separate backend issue)_

Julesssss commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,161 @@
+        const cachedTotal = this.props.iou.cachedTotal.replace(/[()]/g, '');
+        const amount = Number(cachedTotal.substring(1));
+
+        if (amount === 0) {
+            return null;
+        }
```
This is not quite working as expected, we seem to keep showing the preview with an amount of 0 until logging out/in again.

tugbadogan replied:
I'm now using report.hasOutstandingIOU 👍

Julesssss commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,161 @@
+                    <MultipleAvatars
+                        avatarImageURLs={[managerAvatar, ownerAvatar]}
+                        styles={{
+                            secondAvatar: {
```
is it possible to reuse the existing secondAvatar style? Or move this to a new styles object

tugbadogan replied:
Existing secondAvatar style looks like this and I had to use custom style to put avatars in same line as in the issue description. Let me move this to a new styles object.

<img width="568" alt="Screenshot 2021-03-31 at 21 40 35" src="https://user-images.githubusercontent.com/3853003/113208552-d81e0700-9269-11eb-9c63-87ec489bfc70.png">

Julesssss commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -160,4 +182,7 @@ export default withOnyx({
     betas: {
         key: ONYXKEYS.BETAS,
     },
+    report: {
+        key: ({reportID}) => `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+    },
```
If we already have `reportID`, isn't this unnecessary -- now that we simply pass reportID

tugbadogan replied:
This is `chatReportID` but we need `iouReportID` and we need to fetch chatReport first to use `report.iouReportID` in child component's withOnyx. I'm now passing report to use `report.hasOutstandingIOU` anyway.

marcaaron replied:
Can we pass the `iouReportID` or anything else we need directly?

tugbadogan commented:
> 1. Once an IOU is settled we don't need to render the component
>    I think we need to check `isMostRecentIOUReport` AND `report.hasOutstandingIOU`

👍

> 1. If you don't have a reason, we should remove the 'for': (Requested $9 from Jules **for**)

These action messages are generated on server side. Should I modify the message on the client side?

> 1. When a transaction is cancelled, we don't seem to be immediately updating other IOUs.
>    * the middle component should be hidden, (this happens after a refresh)
>    * the preview component should say $5, not $10 -- I can see the `cachedTotal` is correctly showing $5

I fixed these issues 👍
 
> _Also, we should just show a users first name for the requested quote message (but this is a separate backend issue)_

Agreed, this is similar to 'for' issue. It is better to fix this on backend.

tugbadogan committed:
Fixed live update issues and addressing comments

tugbadogan committed:
Fixed live update issues and addressing comments

iwiznia commented:
> These action messages are generated on server side. Should I modify the message on the client side?

No, but after the for there _should be_ the comment of the request, I think that we are not removing the `for` when there's no comment. I will make a PR to handle that

> Also, we should just show a users first name for the requested quote message (but this is a separate backend issue)

Will handle this too.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -307,6 +307,7 @@ function fetchChatReportsByIDs(chatList) {
                 simplifiedReports[reportKey].iouReportID = iouReportObject.reportID;
                 simplifiedReports[reportKey].hasOutstandingIOU = iouReportObject.stateNum === 1
                     && iouReportObject.total !== 0;
+                Onyx.merge(iouReportKey, iouReportObject);
```
Maybe I'm missing something, but shouldn't this be redundant? As the whole simplifiedReports collection is merged to Onyx in line 316/317?

Is mergeCollection not triggering an update? Because if so that sounds like an issue @barun1997 is currently facing too.  I'll take a look into this today.

tugbadogan replied:
I have found that Onyx.mergeCollection doesn't trigger update on individual key subscribers and it was causing delay on updating outstanding amount. Adding Onyx.merge here fixed that issue.

barun1997 replied:
Yup, this sounds exactly like the issue I was facing. But at the same time, I don't think Onyx.merging each individual key is a good fix to the issue since it adds too many renders if I am not wrong, @tugbadogan? Do give me feedback if you've any suggestions so I could implement the same in my PR as well.

Julesssss replied:
Okay, thanks for confirming. I'll get back to you

Julesssss replied:
This was a bug in Onyx, which [has just been fixed](https://github.com/Expensify/react-native-onyx/pull/57). We've just [bumped](https://github.com/Expensify/Expensify.cash/pull/2200) the Onyx version in Expensify.cash, so you should be able to resolve your issue after merging master.

tugbadogan replied:
👍 Thanks for confirming @Julesssss 

I have a question about rebasing to master. It looks like Github requires force push after rebase and in Expensifh.cash contributor guidelines it's recommended not to force push since it ignored previous reviews. 

Do you know if it's possible to rebase and update the PR without force push? or is it OK to force push after rebase?

AndrewGable replied:
I believe in this case it is required to force push, so it is OK. We will discuss internally the changes required to the contributor guidelines for future cases of this. Thank you for double checking first 👍

trjExpensify replied:
Hey @tugbadogan are you all good here now?

tugbadogan replied:
I didn't have a change to look at it today. I'm planning to test it tomorrow. 👍

tugbadogan replied:
I've tested with most recent Onyx version and it seems to be working but I have discovered another problem and created a separate issue for it  https://github.com/Expensify/Expensify.cash/issues/2225

This issue is causing preview box to disappear when it's hovered with mouse. Since `Onyx.mergeCollection` doesn't update report object properly, we don't have `hasOutstandingIOU` value available in this component.

@Julesssss is there any chance your test didn't work as expected due to this issue?


https://user-images.githubusercontent.com/3853003/113597295-6829cb00-9633-11eb-87a9-fc54b29292ad.mov

AndrewGable replied:
Continuing discussion of bug found in linked issue

Julesssss commented:
I ran through the full flow on web, Android, and iOS. I can see the last issues I mentioned have been fixed, but noticed one remaining issue: 

When I pay/decline/create an IOU transaction I see the storage being updated locally. But, we're not updating the current page. I think this is because we're updating the wrong data store. I see the IOUReport being persisted as a regular report under `reports_123`, instead of `reportIOUs_123`.

Julesssss commented:
Just wanted to leave the test flow that I ran though here, in case anyone else tests this over the UK easter bank holiday:

```
1 . Open report with 0 IOUs
2. No preview components should be visible
3. Create an IOU transaction 
4. Preview component should be visible
5. Create a second IOU transaction 
6. A quote message should appear for each, but the preview should only show on the most recent IOU transaction
7. Create a 3rd transaction
8. Cancel a transaction (as the user who created them)
9. Quote should read 'Cancelled'  
10. Decline a transaction (as the user who received them)
11. Quote should read 'Declined'   
12. Pay the IOU
13. Preview should not display
```

### IOU API requests
```
**CreateIOUReport**
_Requesting money from another person_
```
curl --request POST \
  --url 'https://expensify.cash/api?command=CreateIOUTransaction' \
  --header 'Content-Type: multipart/form-data; boundary=---011000010111000001101001' \
  --cookie 'email=admin%2540verified.com; accountID=262; authToken=B2C581...DAE9A' \
  --form debtorEmail=jules+2@expensify.com \
  --form currency=USD \
  --form 'comment=Ice cream' \
  --form amount=299
```

**CreateIOUSplit**
_Splitting bill of $15.00 between myself and two others -- must first create a chat with `CreateChatReport` if one doesn't already exist_
```
curl --request POST \
  --url 'https://expensify.cash/api?command=CreateIOUSplit' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --cookie 'email=jules%2540expensify.com; accountID=2; authToken=4FC46..817617' \
  --data reportID=712 \
  --data amount=1500 \
  --data currency=USD \
  --data 'splits=[{"amount":500,"email":"jules@expensify.com"},{"amount":500,"email":"jules+6@expensify.com"},{"amount":500,"email":"jules+5@expensify.com"}]'
```

**CreateChatReport**
_Must occur prior to CreateIOUSplit if you don't already have the reportID_ 
```
curl --request POST \
  --url 'https://expensify.cash/api?command=CreateChatReport' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --cookie 'email=jules%2540expensify.com; accountID=2; authToken=4FC46...817617' \
  --data 'emailList=["jules+4@expensify.com","jules+6@expensify.com","jules+6@expensify.com","jules+5@expensify.com"]'
```

**Pay IOU**
_Pay outstanding amount to close IOU_
```
curl --request POST \
  --url 'https://expensify.cash/api?command=PayIOU' \
  --header 'Content-Type: multipart/form-data; boundary=---011000010111000001101001' \
  --cookie 'email=admin%2540verified.com; accountID=262; authToken=B2C581...DAE9A' \
  --form reportID=778 \
  --form paymentMethodType=Venmo
```

**Decline IOU**
_Cancel or decline a single IOU transaction/request_
```
curl --request POST \
  --url 'https://expensify.cash/api?command=RejectTransaction' \
  --header 'Content-Type: multipart/form-data; boundary=---011000010111000001101001' \
  --cookie 'email=admin%2540verified.com; accountID=262; authToken=B2C581...DAE9A' \
  --form reportID=776 \
  --form transactionID=4294967349 \
  --form 'comment='\''I dispute this!'\'''
```

tugbadogan commented:
> I ran through the full flow on web, Android, and iOS. I can see the last issues I mentioned have been fixed, but noticed one remaining issue:
> 
> When I pay/decline/create an IOU transaction I see the storage being updated locally. But, we're not updating the current page. I think this is because we're updating the wrong data store. I see the IOUReport being persisted as a regular report under `reports_123`, instead of `reportIOUs_123`.

To clarify, do you expect it to work without page reload? Or are these issues happening even after page reload? I realised that we have live update `ADDCOMMENT` actions but we don't update IOU Actions without page reload. 

I didn't understand which value is updated under `reports_123` data store that we should have stored in `reportIOUs_123`. Am I expected to call API and update IOU report data when there is a change in regular reports stored under `reports_123`?

iwiznia commented:
> To clarify, do you expect it to work without page reload? 

I am unsure I understand the issue, but yeah, everything should work without page reloads

>  I realised that we have live update ADDCOMMENT actions but we don't update IOU Actions without page reload.

Are you saying that we are not sending push notifications for IOU actions?

tugbadogan committed:
Update chat report real time on IOU report changes

tugbadogan committed:
Merge branch 'tugbadogan-create-iou-preview-component' of https://github.com/tugbadogan/Expensify.cash into tugbadogan-create-iou-preview-component

tugbadogan committed:
Fix linter errors

shawnborton commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,162 @@
+        const cachedTotal = this.props.iou.cachedTotal.replace(/[()]/g, '');
+
+        return (
+            <View style={[styles.iouPreviewBox, {width: this.props.isSmallScreenWidth ? '100%' : '50%'}]}>
```
Could we have this use a width of 100%, but give it a max-width of something like `300` or something? This should simplify this statement here.

shawnborton commented on src/styles/styles.js:
```diff
@@ -1256,6 +1267,18 @@ const styles = {
+        borderColor: themeColors.border,
+        borderWidth: 1,
+        borderRadius: variables.componentBorderRadiusNormal,
+        padding: 15,
```
Let's use 20 for padding

shawnborton commented on src/styles/styles.js:
```diff
@@ -1256,6 +1267,18 @@ const styles = {
+        borderWidth: 1,
+        borderRadius: variables.componentBorderRadiusNormal,
+        padding: 15,
+        marginTop: 15,
```
Let's use 16 for this please.

shawnborton commented on src/styles/styles.js:
```diff
@@ -1256,6 +1267,18 @@ const styles = {
+        marginTop: 15,
+    },
+
+    normalRadius: {
```
I'm not following exactly why we need this. Border radius should just be part of how we define the button style, not an extra style we need to add.

shawnborton commented on src/styles/styles.js:
```diff
@@ -1256,6 +1267,18 @@ const styles = {
+    iouPreviewBox: {
+        borderColor: themeColors.border,
+        borderWidth: 1,
+        borderRadius: variables.componentBorderRadiusNormal,
```
Could we add a new variable, something like `variables.componentBorderRadiusCard` and set it to `12` and then use it here? I think we want all card-like components to have a slightly bigger border radius than normal.

shawnborton commented:
Looking great so far, left a couple of style comments for you. Another thing I noticed is that the two overlapping avatars don't seem to be far enough over to the right edge of the IOU preview box.

tugbadogan commented:
> > To clarify, do you expect it to work without page reload?
> 
> I am unsure I understand the issue, but yeah, everything should work without page reloads
> 
> > I realised that we have live update ADDCOMMENT actions but we don't update IOU Actions without page reload.
> 
> Are you saying that we are not sending push notifications for IOU actions?

I was confused as some push notifications are going to IOU report ID (when IOU is created) and some of them are going to the chat report ID (when IOU is updated). I had to handle them separately.


I modified Report.js to handle IOU actions and update corresponding chat report ID by finding chat report ID via participants as we don't have ChatReportID yet when IOU is created.


The page and preview component are now updated immediately on create/pay/cancel actions without page reload.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -265,6 +265,29 @@ function updateIOUReportData(chatReport) {
+ */
+function findChatReportID(participants) {
+    if (participants.length < 2) {
+        throw new Error('fetchOrCreateChatReport() must have at least two participants');
```
The error message should be updated to say 'findChatReportID() must have at least two participants'.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -293,8 +316,23 @@ function fetchChatReportsByIDs(chatList) {
+                    findChatReportID([report.ownerEmail, report.managerEmail]).then((chatReportID) => {
+                        const iouReportKey = `${ONYXKEYS.COLLECTION.REPORT_IOUS}${report.reportID}`;
+                        const reportKey = `${ONYXKEYS.COLLECTION.REPORT}${chatReportID}`;
+                        Onyx.merge(iouReportKey, getSimplifiedIOUReport(report, chatReportID));
+                        const chatReportObject = {};
+                        chatReportObject.iouReportID = report.reportID;
+                        chatReportObject.hasOutstandingIOU = report.stateNum === 1
+                            && report.total !== 0;
+                        Onyx.merge(reportKey, chatReportObject);
```
Don't think we should be merging to Onyx within a loop here. Couldn't we follow the pattern from the first if condition, and collect all the reports in `reportIOUData` and then use mergeCollection on line 354 to merge once?

tugbadogan replied:
Since we are using `findChatReportID` here and it returns a Promise. I don't think we can follow the same pattern here without refactoring the whole function as we cannot await promise synchronously. Is there a better way to handle this part without using `Promise` or `Onyx.merge`?

I think this would be OK because we call fetchChatReportsByIDs with multiple chat report IDs or a single IOU report ID. And this merge action is called only for that single IOU report ID.

Julesssss commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,162 @@
+        const managerAvatar = lodashGet(this.props.personalDetails, [this.props.iou.managerEmail, 'avatar'], '');
+        const ownerAvatar = lodashGet(this.props.personalDetails, [this.props.iou.ownerEmail, 'avatar'], '');
+        const sessionEmail = lodashGet(this.props.session, 'email', null);
+        const cachedTotal = this.props.iou.cachedTotal.replace(/[()]/g, '');
```
Not sure what exactly causes this, but I'm seeing an occasional crash here. Could you add some protection for `this.props.iou.cachedTotal` here please:
![Screenshot 2021-04-06 at 14 31 32](https://user-images.githubusercontent.com/10736861/113718838-f3af6480-96e4-11eb-8c7a-2dddf25c5da1.png)

Julesssss commented:
I think we're hiding the Preview component in some cases when we shouldn't be: 

- Login to user with outstanding IOUs
- Click into a chat with outstanding IOU
- IOU Preview will display as expected
- Refresh page
- IOU Preview **will not** display as expected


https://user-images.githubusercontent.com/10736861/113721364-5f92cc80-96e7-11eb-93e0-a51335359679.mov

Julesssss commented:
I know you didn't touch the `ReportActionCompose` class, but I've had a crash a couple of times due to the report prop being null when an IOU change takes place (pay an IOU, and then send a new transaction).

Could you please add a default prop:  `report: {},` [here](https://github.com/Expensify/Expensify.cash/blob/5551c434f5a43d344d6385a0469eeecb796f6ba7/src/pages/home/report/ReportActionCompose.js#L44)? This should prevent a crash and help with testing.

Julesssss reviewed (changes requested):
Thanks for making the last batch of changes, it's looking really good. I left a few additional issues and change requests.

tugbadogan commented:
> I think we're hiding the Preview component in some cases when we shouldn't be: 
> 
> 
> 
> - Login to user with outstanding IOUs
> 
> - Click into a chat with outstanding IOU
> 
> - IOU Preview will display as expected
> 
> - Refresh page
> 
> - IOU Preview **will not** display as expected
> 
> 
> 
> 
> 
> https://user-images.githubusercontent.com/10736861/113721364-5f92cc80-96e7-11eb-93e0-a51335359679.mov
> 
> 
> 
> 

Yes, I'm aware of this problem. This issue is caused by the following issue I have created.

https://github.com/Expensify/Expensify.cash/issues/2225

Julesssss commented:
> Yes, I'm aware of this problem. This issue is caused by the following issue I have created.
> 
> #2225

Ah I missed that, thank you!

Julesssss commented:
Hey @tugbadogan, we've [just merged the fix](https://github.com/Expensify/Expensify.cash/issues/2225) to your Onyx test.

tugbadogan commented:
Thanks for the notice @Julesssss!

I was planning to fix styling changes that @shawnborton recommended before I rebase on master, force push and fix ReportActionCompose.

AndrewGable commented:
That sounds good!

trjExpensify commented:
Hey @tugbadogan! Can we get an update on progress here, please? We’re ideally looking to close this issue out by the end of this week, so we can focus on the remaining pieces of the project next week. Thanks!

tugbadogan commented:
@trjExpensify I'm working on it now and will update the PR in a couple of hours. 👍

tugbadogan committed:
Addressing comments and styling changes

tugbadogan committed:
Merge branch 'master' into tugbadogan-create-iou-preview-component

tugbadogan committed:
Fixing merge issues

tugbadogan commented:
I updated the PR and addressed all of your comments.

shawnborton commented:
Could you please update the screenshots as well? Thanks!

tugbadogan commented:
@Julesssss @marcaaron I tested with the most recent Onyx version but I think mergeCollection still has some problems. Here is how to reproduce the problem:

- Create 2 IOU transactions
- Cancel one of them
- Preview component doesn't update even though we have reduced cachedTotal amount in localStorage
- Used Onyx.merge instead of Onyx.mergeCollection 
```
Onyx.merge(reportKey, simplifiedReports[reportKey]);
Onyx.merge(iouReportKey, reportIOUData[iouReportKey]);
```
- Retry step 1-2 and observe that preview component shows correct value.

tugbadogan commented:
@shawnborton I updated the screenshots 👍

marcaaron commented:
>I think mergeCollection still has some problems

Can you explain exactly what are you using it for? In most cases, we should be using `merge()` and only `mergeCollection()` when setting many keys of data at once.

tugbadogan committed:
Create a separate function for updateIOUAndChatReportData

tugbadogan commented:
> > I think mergeCollection still has some problems
> 
> Can you explain exactly what are you using it for? In most cases, we should be using `merge()` and only `mergeCollection()` when setting many keys of data at once.

I was reusing fetchChatReportsByIDs to link IOU and chat reports together. And that function is already using `mergeCollection`
https://github.com/Expensify/Expensify.cash/blob/3dfea0090410aed64199f2e37680429f2682ee0a/src/libs/actions/Report.js#L316-L317

I've update the PR and created a separate function that takes single reportID and find corresponding chat/IOU report object and link them together.

This resolved the issues with `mergeCollection` and we don't have to call `Onyx.merge` in a loop now @Julesssss

tugbadogan commented:
By the way, I started getting following exception from the API, I guess my latest IOU report is malformed on the server side.

http://localhost:8080/api?command=Report_GetHistory
reportID: 70001894

`{"code":666,"jsonCode":404,"type":"Expensify\\Error\\Transaction\\NoLongerExists","UUID":"B5B9CEC4-981E-4D81-9C8B-7CAB34587944","message":"Transaction no longer exists","title":"","data":{"transactionID":52804612859625595},"htmlMessage":"","requestID":"095553a6bb0000546a8a9c5000000001"}`

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -869,7 +971,12 @@ function handleReportChanged(report) {
+    }
+
+    // If this is a brand new IOU report, find and link corresponding chat report ID
+    if (report.reportName === CONST.REPORT.ACTIONS.TYPE.IOU && report.chatReportID === undefined) {
```
Hmm so there are two ways to tell if a report is an IOU? Both the title and the `rNVP` - can we maybe pick one? This feels kind of like implementation details that we shouldn't have to worry about here.

tugbadogan replied:
We don't have rNVP in this report object. I removed this part anyway since it's not needed anymore.

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +349,77 @@ function fetchChatReportsByIDs(chatList) {
+    if (!iouReportObject.participants) {
+        return;
+    }
+    findChatReportID([iouReportObject.participants[0], currentUserEmail])
```
I don't get why we are not first looking at the reports we have to see if the one belonging to the `iouReportObject` is already in storage.

tugbadogan replied:
This function is not used anymore, so I deleted it.

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +349,77 @@ function fetchChatReportsByIDs(chatList) {
+                    const reportKey = `${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`;
+                    Onyx.merge(reportKey, chatReportObject);
+                });
+            } else if (lodashGet(report, 'reportNameValuePairs.type') === 'iou') {
```
This should, in my opinion, be separated into two methods each one handling the different report type.

Julesssss commented on src/components/ReportActionItemIOUQuote.js:
```diff
@@ -0,0 +1,29 @@
+    <View style={[styles.chatItemMessage]}>
+        {_.map(_.compact(action.message), (fragment, index) => {
+            const viewDetails = '<br /><a href="#">View Details</a>';
+            const html = `<blockquote>${fragment.text}${viewDetails}</blockquote>`;
```
This was left simple to keep making progress with IOU. I'll be replacing it with an actual component in the IOU Details issue so that I can link to a modal when clicking 'View Details'

marcaaron replied:
Ok. My concern is that someone else might see this and think it's correct, but if there are plans to fix it quickly maybe that's fine.

shawnborton replied:
So the styling of this doesn't quite match our mockups:

![image](https://user-images.githubusercontent.com/2319350/114237367-d920f980-9937-11eb-97bf-0a48d7f14d61.png)


versus

![image](https://user-images.githubusercontent.com/2319350/114237429-f1911400-9937-11eb-8507-5a3ec0913d25.png)

Is this the PR where I should share my feedback on how to accomplish the mockups, or will this be tackled elsewhere?

tugbadogan replied:
Hi @shawnborton 

I didn’t update the styles of HTML tags since the plan is to replace HTML part with standalone components later.

Julesssss replied:
Hey @shawnborton. 

Let's keep that feedback for the new PR which will replace the HTML. I'll make sure to tag you in the PR.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -265,6 +265,29 @@ function updateIOUReportData(chatReport) {
+            }
+            return data.reportID;
+        });
+}
```
Let's simplify this to have one function for returning the report. The existing `fetchOrCreateChatReport` can then call the first function, persist to Onyx and then be renamed to something more descriptive.

tugbadogan replied:
I modified `fetchOrCreateChatReport` and reused `findChatReportID` function. I renamed `fetchOrCreateChatReport` as `fetchOrCreateChatReportAndRedirect` to make it more descriptive.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -293,8 +316,11 @@ function fetchChatReportsByIDs(chatList) {
             _.each(fetchedReports, (report) => {
-                const simplifiedReport = getSimplifiedReportObject(report);
-                simplifiedReports[`${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`] = simplifiedReport;
+                // Do not store IOU reports as 'report_' Onyx keys.
+                if (lodashGet(report, 'reportNameValuePairs.type') === 'chat') {
+                    const simplifiedReport = getSimplifiedReportObject(report);
+                    simplifiedReports[`${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`] = simplifiedReport;
+                }
```
Yeah we shouldn't be passing in IOU reports to `fetchChatReportsByIDs` in the first place.

Just tested the response to `Get&returnValueList=chatList`, and all of these reports are of `type: 'chat'` as we should expect. So, either this new logic is unnecessary, or something else is causing IOU reports to slip through in the response. 

@tugbadogan any idea if this is the case?

tugbadogan replied:
It looks like we don't need this check anymore as I created a separate function named `updateIOUAndChatReportData`.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -869,7 +971,12 @@ function handleReportChanged(report) {
     // and the report does not yet exist in Onyx (eg. a new DM created with the logged in person)
     if (report.reportID && report.reportName === undefined) {
-        fetchChatReportsByIDs([report.reportID]);
+        updateIOUAndChatReportData(report.reportID);
```
Yeah, we should solve the problem of IOU reports being seen as chat reports at the place it occurs, rather than filtering them out here.

tugbadogan replied:
When a new IOU report is created we receive a push notification for CREATED action with IOU report ID. 
We call updateReportWithNewAction and set Onyx object for report.

When we get a push notification with CREATED action, we don't know if it's an IOU or Chat report until we call API to fetch reportStuff.


It would be great if we fix the push notifications in the first place, so we don't have to reuse chatReport related functions and put if conditions.

Julesssss replied:
Ah, thanks for pointing out the problem. I looked into the data to confirm and it seems that we are returning the wrong action data (we're sending the very first action, instead of the latest IOU action).

Sorry that we didn't notice this earlier. I have created an internal issue so that we can unblock you asap. Perhaps there is a workaround via looking up the report after the notification is received, but I would like to avoid this and focus on fixing the actual issue.

tugbadogan replied:
There is another push notification issue btw. When an IOU report is created, we don't get a new action push for the chat report, so we cannot display "Requested £.... for ...." quoted message until the page is refreshed.

Julesssss replied:
I'll look into that also. Thanks.

Julesssss replied:
Hi @tugbadogan, we've fixed both of these issues but it will take a couple of days to deploy to production. I'll let you know as soon as this has happened.

Julesssss replied:
Hey again @tugbadogan. Ignore my last message, the fix should now be live on production 🎉 

Please let us know if you have any further issues here.

tugbadogan replied:
It looks like we are getting push notifications for just chat reports 👍 I updated PR and separated IOU update functions from chat report functions. We are not creating `report_<>` for IOU reports and we don't have to modify `handleReportChanged` function anymore.

marcaaron commented on src/pages/home/report/ReportActionItemGrouped.js:
```diff
@@ -4,16 +4,35 @@ import PropTypes from 'prop-types';
     <View style={[styles.chatItem]}>
         <View style={[styles.chatItemRightGrouped]}>
-            <ReportActionItemMessage action={action} />
+            {action.actionName === 'IOU'
```
NAB, I think at this point this component has probably outgrown it's original design and should be refactored eventually.

e.g. 

We have a group view and ungrouped view and instead of adding two checks in both components to figure out what content to render we probably should be passing in the message content as a child from a parent. This way we do not need to have duplicate logic in two components as if anything about them changes in the future we must remember to update both the grouped and ungrouped case.

tugbadogan replied:
I refactored this part.

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -278,6 +278,21 @@ class ReportActionsView extends React.Component {
+            .max(action => action.sequenceNumber)
+            .value().sequenceNumber;
+        return actionSequenceNumber === mostRecentIOUReportSequenceNumber;
+    }
```
It seems like we are performing this loop as a way of "flagging" an action as the `mostRecentIOUReportSequenceNumber`. This seems inefficient to me since we have `n` number of actions and only ONE of them will be most recent. Yet each time we render an action we will loop over all to find the number again and compare it to the action's `sequenceNumber`.

Here are some alternative ideas...

**Store this on the report instead**

- When updating a report's actions look for the most recent IOU action and set the sequenceNumber to the report object
- When updating the actions we'll update this number.

**Only calculate when the actions change**
 
Only calculate the `mostRecentSequenceNumber` when the actions change in the `updateSortedReportActions` method here:

https://github.com/Expensify/Expensify.cash/blob/3dfea0090410aed64199f2e37680429f2682ee0a/src/pages/home/report/ReportActionsView.js#L196-L203

shawnborton commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,160 @@
+    createPayButton() {
+        return (
+            <TouchableOpacity
+                style={[styles.buttonSmall, styles.normalRadius, styles.buttonSuccess, styles.mt6]}
```
Could you try using `styles.mt4` so we reduce the top margin just a little bit?

shawnborton replied:
Also, is there a reason why we have `styles.normalRadius` on here? I think the radius should just be coming from `styles.buttonSmall`

tugbadogan replied:
I forgot to remove it from here. I updated the code.

shawnborton commented on src/styles/styles.js:
```diff
@@ -1255,6 +1267,20 @@ const styles = {
+    },
+
+    iouPreviewBoxAvatar: {
+        marginRight: -10,
```
Is there a way to implement this component without needing negative margin? Ideally the view that wraps the two avatars will float all the way to the right by default.

tugbadogan replied:
`MultipleAvatars` has `marginRight: 16` and uses absolute to position avatars. We have to rewrite MultipleAvatars component to achieve that result without using negative margin here.

Do you recommend modifying `MultipleAvatars` to accept style prop where we can set `marginRight: 0` and `width: 46` instead of `marginRight: 16` and `width: 40`?

shawnborton replied:
Ah that's fair, I didn't realize we were re-using an existing component. I think what you have is fine then.

shawnborton commented on src/styles/styles.js:
```diff
@@ -1255,6 +1267,20 @@ const styles = {
+        borderRadius: variables.componentBorderRadiusCard,
+        padding: 20,
+        marginTop: 16,
+        maxWidth: 400,
```
Could you change this to 300 instead?

tugbadogan commented on src/libs/actions/Report.js:
```diff
@@ -416,6 +513,11 @@ function updateReportWithNewAction(reportID, reportAction) {
 
+    // If chat report receives an action with IOU, update IOU object
+    if (reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.IOU) {
+        updateIOUAndChatReportData(reportID);
```
With push notification, we're just getting an update about a transaction or there's a new IOU action for this chat report ID. We don't have enough data to update IOU data, that's why we need to make another network request to fetch updated IOU data that has the correct amount.

tugbadogan committed:
Addressing comments

tugbadogan committed:
Merge branch 'master' into tugbadogan-create-iou-preview-component

tugbadogan commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -278,6 +278,21 @@ class ReportActionsView extends React.Component {
+     * @param {Number} actionSequenceNumber
+     * @returns {Boolean}
+     */
+    isMostRecentIOUReport(actionSequenceNumber) {
```
Renamed as isMostRecentIOUReportAction.

tugbadogan commented:
Thanks for all the review. I updated the PR.

Julesssss commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,135 @@
+    personalDetails,
+    session,
+}) => {
+    const managerName = lodashGet(personalDetails, [iou.managerEmail, 'displayName'], iou.managerEmail);
```
Not sure where the best place for this would be, but could you add a small comment that manager email is the current owner of the report? So that the `iou.managerEmail === sessionEmail` makes more sense?

tugbadogan replied:
I added comment and used `Str.removeSMSDomain` here.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -580,26 +684,14 @@ function unsubscribeFromReportChannel(reportID) {
-            if (data.jsonCode !== 200) {
-                throw new Error(data.message);
-            }
+function fetchOrCreateChatReportAndRedirect(participants) {
```
NAB: `fetchOrCreateChatReportAndNavigate` is a bit more platform agnostic.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -265,6 +271,29 @@ function updateIOUReportData(chatReport) {
 }
 
+/**
+ * Get the report ID for a chat report for a specific
+ * set of participants.
+ *
```
I don't think a line break is needed here

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +352,76 @@ function fetchChatReportsByIDs(chatList) {
+ *
+ * @param {Object} report
+ */
+function updateChatReportDataForIOUReport(report) {
```
These IOU<>CHAT report functions can be a bit confusing. Perhaps naming the param `chatReport` might slightly help here.

tugbadogan replied:
Renamed it as `iouReport`.

Julesssss commented:
Not quite there yet, but I think the latest changes were a positive improvement.

The contributor mentioned an issue that we had not yet discovered: IOU pusher/push notification data is not correct and we seem to be sending the first action from the report, rather than the actual IOU action. 

Created a daily issue here and have shared with engineering chat to get someone to look pick it up ASAP., will bump again later if nobody has taken it. As further contributor changes would be necessary once fixed, this should be the IOU project priority issue.

Julesssss commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -80,15 +103,33 @@ class ReportActionItem extends Component {
-                                    ? <ReportActionItemSingle action={this.props.action} />
-                                    : <ReportActionItemGrouped action={this.props.action} />}
+                                    ? (
+                                        <ReportActionItemSingle
+                                            action={this.props.action}
+                                            message={message}
+                                        />
+                                    )
+                                    : (
+                                        <ReportActionItemGrouped
+                                            message={message}
+                                        />
```
Is action missing from `ReportActionItemGrouped`, or is action unnecessary in `ReportActionItemSingle`?

tugbadogan replied:
`action` is used for other purposes in `ReportActionItemSingle`. (It is used to show avatar of action's actor.)

NikkiWines commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -13,6 +15,9 @@ import PopoverWithMeasuredContent from '../../../components/PopoverWithMeasuredC
 import ReportActionContextMenu from './ReportActionContextMenu';
+import ReportActionItemIOUPreview from '../../../components/ReportActionItemIOUPreview';
+import ReportActionItemMessage from './ReportActionItemMessage';
+
```
NAB: extra line here

tugbadogan replied:
Removed the extra line

NikkiWines reviewed (dismissed):
Couple of minor comments

tugbadogan committed:
Updating Report.js to work with fixed push notifications

tugbadogan committed:
Merge pull request #14 from Expensify/master

Merge

tugbadogan committed:
Merge branch 'master' into tugbadogan-create-iou-preview-component

tugbadogan commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,135 @@
+            <ReportActionItemIOUQuote action={action} />
+            {isMostRecentIOUReportAction
+                    && report.hasOutstandingIOU
+                    && iou ? (
```
Started using `&&` here 👍

tugbadogan commented:
Updated the PR.

Julesssss commented:
Ran through the test flow after merging master again locally (for the full IOU Modal flow). Working nicely. One thing I noticed is that we don't seem to be sending push notifications for each of the individually created 1:1 IOUs when creating a bill split, but only the group message notification which I think we'll need to fix. I'll raise an issue internally for this.

tugbadogan commented:
@Julesssss Is there anything else I need to do here?

Julesssss commented:
> @Julesssss Is there anything else I need to do here?

Not for the moment. I think it's best to test the flow with functioning group notifications. Hopefully we can resolve this quickly.

Julesssss commented:
Unless others have additional review feedback? Perhaps now is a good time to review the current code while we're awaiting fixes to let us test the full bill split flow.

@NikkiWines @marcaaron @shawnborton @AndrewGable would you mind re-reviewing?

tugbadogan committed:
Merge pull request #15 from Expensify/master

Merge

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -576,30 +685,18 @@ function unsubscribeFromReportChannel(reportID) {
-    if (participants.length < 2) {
-        throw new Error('fetchOrCreateChatReport() must have at least two participants');
-    }
+function fetchOrCreateChatReportAndNavigate(participants) {
```
I think the impact of this change can be significantly lessened by just adding a parameter indicating we don't want to navigate. Can we try that instead?

```js
function fetchOrCreateChatReport(participants, shouldNavigate = true) {
...
}
```

Then just pass false wherever you don't need to navigate ? Or vice-a-versa..

marcaaron commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -50,6 +70,8 @@ class ReportActionItem extends Component {
         return this.state.isPopoverVisible !== nextState.isPopoverVisible
             || this.props.displayAsGroup !== nextProps.displayAsGroup
+            || this.props.isMostRecentIOUReportAction !== nextProps.isMostRecentIOUReportAction
+            || !_.isEqual(this.props.report, nextProps.report)
```
I haven't benchmarked this but subscribing to the report object in _every_ message seems potentially bad for performance since reports data changes frequently.

marcaaron commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -95,8 +126,17 @@ class ReportActionItem extends Component {
+                                    )
+                                    : (
+                                        <ReportActionItemGrouped
+                                            message={message}
```
Could we maybe use `children` here instead of `message` (and also in `ReportActionItemSingle`)?

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -232,6 +234,19 @@ class ReportActionsView extends React.Component {
+    /**
+     * Finds and updates most recent IOU report action number
+     *
+     * @param {Array<{sequenceNumber, actionName}>} reportActions
```
Bad doc as it doesn't return an `Array`

tugbadogan replied:
This is for function `param` not `return`. Am I missing something here?

marcaaron replied:
Nope, just my brain on a Friday misread this.

marcaaron reviewed (changes requested):
I'm not sure what other changes we are waiting on at this point, but I have another round of comments. I'd love for us to focus on cleaning up the names and logic in `Report.js` some more.

NikkiWines reviewed (commented):
Aside from Marc's comments, I think this is looking really good. I do agree some of the function naming in Report.js is a bit confusing.

tugbadogan commented on src/components/ReportActionItemIOUPreview.js:
```diff
@@ -0,0 +1,147 @@
+};
+
+const defaultProps = {
+    iou: {},
```
When there is a new IOU action, we don't have IOU report immediately and this component throws exceptions if we mark it as required.

marcaaron replied:
Hmm ok, I'm having trouble understanding the situation where we would have one without the other but OK.

tugbadogan replied:
Also, when we don't have active IOU report, previous IOU actions in the chat report will load this component and we won't have IOU report from Onyx.

marcaaron replied:
Sorry, I don't understand, but it's not a real blocker.

tugbadogan commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +351,82 @@ function fetchChatReportsByIDs(chatList) {
+ * @param {Number} iouReportID
+ * @param {Number} chatReportID
+ */
+function updateIOUReportData(iouReportID, chatReportID) {
```
There is already a function named `fetchIOUReportData` which fetches the data but doesn't update stored data. It is used by `fetchChatReportsByIDs` function to update all chat and IOU reports in batch.

marcaaron replied:
Ok, but why do we need two separate methods that both call the same APIs? Looking at that method it calls both `GetIOUReport` and `Get: reportStuff`. But it should be calling `fetchIOUReport()` and `fetchIOUReportID()` if we also have methods that only do that.

tugbadogan replied:
Now we have 4 functions related to iou reports:
- `fetchIOUReport` takes `iouReportID`, `chatReportID` and calls `Get: reportStuff` and returns `iouReportObject`.
- `fetchIOUReportID` takes `debtorEmail` and calls `GetIOUReport` and returns `iouReportID`.

- `fetchIOUReportDataForChatReport` uses `fetchIOUReportID` and `fetchIOUReport` to return `iouReportObject` for given chat report.
- `updateIOUReportData` takes `iouReportID`, `chatReportID` and  calls `fetchIOUReport` to get `iouReportObject` and updates Onyx data.

marcaaron replied:
I'm confused it looks like we got rid of `fetchIOUReportDataForChatReport`?

marcaaron replied:
The names need to be fixed so they make sense and are easy to understand.

tugbadogan replied:
I just submitted the commit with these 4 functions. Which one is not clear and should be renamed?

marcaaron replied:
`fetchIOUReportDataForChatReport` - doesn't sound very different from `fetchIOUReport()` because all of them are "for chat reports". The method is also only used once so I think there's no reason to create it. For now, could we please just get rid of this method and move the logic here instead to avoid confusion:

https://github.com/Expensify/Expensify.cash/blob/103e1d1199032986013995515dede56c02a08f31/src/libs/actions/Report.js#L312

`updateIOUReportData` is inaccurate because we are _fetching_ data and setting it not updating it. It's not really conventional (for this codebase) to have a method `fetchIOUReport()` that fetches, but doesn't set any data (all of the other actions are fetching and setting data in Onyx). And then have another method `updateIOUReportData()` that both fetches and sets the data.

So, my only suggestion is to remove the calls to `fetchIOUReport()` inside `updateIOUReportData()` and call it before each usage then rename `updateIOUReportData()` to `setLocalIOUReportData()` so it is clear that it will only save the data to Onyx.

tugbadogan replied:
I updated function names as you recommended. Can you check again?

tugbadogan commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +351,82 @@ function fetchChatReportsByIDs(chatList) {
+    if (iouReportID) {
+        updateIOUReportData(iouReportID, chatReportID);
+        return;
+    }
```
This function takes `chatReportID` parameter, not `iouReportID`. So when we are calling this function we don't know if we have `iouReportID` or not for given `chatReportID`.

tugbadogan committed:
Addressing comments

tugbadogan commented:
@marcaaron @NikkiWines I updated the PR based on your comments. Can you review again?

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -416,6 +501,17 @@ function updateReportWithNewAction(reportID, reportAction) {
 
     Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`, reportActionsToMerge);
 
+    // If chat report receives an action with IOU, update IOU object
+    if (reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.IOU) {
+        const chatReport = lodashGet(allReports, reportID);
+        const iouReportID = lodashGet(chatReport, 'iouReportID');
+        if (iouReportID) {
+            updateIOUReportData(iouReportID, reportID);
+        } else {
+            fetchIOUReportID(reportID).then(iouID => updateIOUReportData(iouID, reportID));
+        }
```
Group chat reports will meet the `reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.IOU` condition, but we should exclude them, as they cannot have an associated IOUReport.

tugbadogan replied:
I'm now checking length of participants and don't call the functions for group chats.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +333,81 @@ function fetchChatReportsByIDs(chatList) {
+function fetchIOUReportID(chatReportID) {
+    const chatReport = lodashGet(allReports, chatReportID);
+    const participants = chatReport.participants;
+    if (participants.length !== 1) {
+        console.error('chatReport doesn\'t have any participant.');
+        return;
+    }
```
Following up from my previous comment, group reports will hit this line but should not be seen as an error. 

Lets add another condition to catch reports with multiple participants and simply return.

Julesssss commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +333,81 @@ function fetchChatReportsByIDs(chatList) {
+ */
+function fetchIOUReportID(chatReportID) {
+    const chatReport = lodashGet(allReports, chatReportID);
+    const participants = chatReport.participants;
```
Noticed one more bug here. In the case where an IOUAction is sent between two users who have not previously sent chat comments, then chatReport is undefined and causes an error here. 

- Create/find two new users who have not communicated with each other
- Create an IOU Request from A to B
- User B receives the push, but [chatReport is null here](https://github.com/Expensify/Expensify.cash/blob/975c0fc4bd158de51cf9e3c8e3fee2d91c5a81af/src/libs/actions/Report.js#L387)
- Unable to continue to get IOUReport, error logged

Thankfully it should be possible to get the name of the sender via `reportActionsToMerge.lastActorEmail` data [persisted to Onyx here](https://github.com/Expensify/Expensify.cash/blob/main/src/libs/actions/Report.js#L417). 

![Screenshot 2021-04-19 at 15 26 22](https://user-images.githubusercontent.com/10736861/115253503-590f4680-a124-11eb-94b0-52e65e6c2332.png)

Julesssss replied:
Actually, as `fetchIOUReportID` is only called if the report is not available locally, I think we can modify the function to always take `reportAction.actorEmail` as a parameter (and update the function name).

Pleas let me know if you have any further questions!

tugbadogan replied:
I updated function to get debtor email as a parameter.

tugbadogan replied:
This method works for user B. But when user A receives an update, we need to pass user B's email address. But report action doesn't have user B's email address here. What do you recommend when current user is the creator of fresh IOU report?

Julesssss replied:
Okay, it seems like adding the IOUReportID to the push notification data would save us a lot of trouble here. Let me take a look and get back to you. Perhaps I can resolve this in the follow up issue.

Julesssss replied:
Okay, so no need to worry about this issue. 

I'll use `pushJSON.originalMessage.transactionID` to retrieve the IOUReportID, using a [function that is already defined here](https://github.com/Expensify/Expensify.cash/blob/ccf9b7f0b158b4ae7d7ef3831f861018139aab90/src/libs/actions/IOU.js#L112-L131).

marcaaron commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -58,6 +58,7 @@ const propTypes = {
 const defaultProps = {
     comment: '',
     modal: {},
+    report: {},
```
What is this change related to? I would expect the report to be available if we are loading a chat for one, did we encounter some error?

tugbadogan replied:
@Julesssss told me this was throwing exceptions.

Julesssss replied:
I'm afraid I don't recall where this was causing a problem, but I had to add this locally to prevent a repeated error.

Julesssss replied:
Ah found it: the following error occurs when receiving an IOU Report notification:

![Screenshot 2021-04-20 at 12 02 37](https://user-images.githubusercontent.com/10736861/115385880-93362200-a1d0-11eb-9db8-89641560cc0a.png)

marcaaron reviewed (changes requested):
Couple more comments.

Julesssss commented:
Finally, would you mind merging `main` (previously master) again. This will make it much simpler to test, because the complete IOU Modal flow is available.

tugbadogan committed:
Merge pull request #16 from Expensify/main

Merge

tugbadogan committed:
Merge branch 'master' into tugbadogan-create-iou-preview-component

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +333,81 @@ function fetchChatReportsByIDs(chatList) {
+            console.debug('GetIOUReport returned a reportID of 0, not fetching IOU report data');
+            return;
+        }
+        return iouReportID;
```
Ok now that we have this can we use it to replace these lines? The code is very similar if not the same?

https://github.com/Expensify/Expensify.cash/blob/f64e2f3b95ba15666f6963ae42cce2fe537e2719/src/libs/actions/Report.js#L235-L249

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -323,6 +333,81 @@ function fetchChatReportsByIDs(chatList) {
+            console.error(`No iouReportData found for reportID ${iouReportID}`);
+            return;
+        }
+        const iouReportObject = getSimplifiedIOUReport(iouReportData, chatReportID);
```
We are duplicating again... 

https://github.com/Expensify/Expensify.cash/blob/f64e2f3b95ba15666f6963ae42cce2fe537e2719/src/libs/actions/Report.js#L251-L272

Mostly the same code.

tugbadogan replied:
I just submitted the commit for these changes.

tugbadogan committed:
Simplify Report.js

tugbadogan committed:
Another set of refactoring in Report.js

tugbadogan committed:
Merge pull request #17 from Expensify/main

Merge

tugbadogan committed:
Merge branch 'master' into tugbadogan-create-iou-preview-component

tugbadogan committed:
Fixed lint error

tugbadogan commented:
Updated the PR.

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -44,6 +44,16 @@ Onyx.connect({
+            allReports[val.reportID] = val;
+        }
+    },
+});
```
NAB, I think we have avoided storing all the report data because of an overhead to storing all reports in memory. Not sure if it matters or not, but we are only accessing the report to get the `participants` so it might make more sense to store only what we need instead of everything.

Julesssss replied:
I can take a look at this once separately. 

Presumable we would switch to a reportParticipants object, with reportID as the key and participants as the value?

marcaaron replied:
Yep that's what I was thinking!

Julesssss replied:
Resolving, will implement this change in the [IOU Details PR](https://github.com/Expensify/Expensify/issues/158811)

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -284,7 +281,29 @@ function fetchChatReportsByIDs(chatList) {
+                    return;
+                }
+
+                // If we don't have one participant (other than the current user), this is not an IOU
```
NAB, this comment would be better if it explained this a bit more. It's unclear how we can even reach this code.

Julesssss replied:
```
// If we don't have one participant (other than the current user), this is not an IOU
should be: 
// Group chat reports cannot and should not be associated with a specific IOU report
```

Also, I don't think we should log an alert message, because this is expected to happen. Could you simply return here?

Julesssss replied:
Or throw an error if participant.length is 0, but simple return if it is 2.

marcaaron replied:
Why would there be `0` participants? Feels like we are optimizing for a case that we may never run into.

Julesssss replied:
I've seen `0 participants` this multiple times when testing today. It's a bug which I haven't yet been able to reliably reproduce/document. It's possibly related to the issue I mentioned [here](https://github.com/Expensify/Expensify.cash/pull/2112#discussion_r615908160), but keeping the alert for now would be helpful.

Julesssss replied:
I would be happy to remove this case once the bug is resolved though.

tugbadogan replied:
I updated this section to make it better.

marcaaron reviewed (dismissed):
Thanks for the changes. Looks much better 👍

Julesssss commented:
I think the Podfile is out of sync since the last `main` merge. Could you run this again and commit the Podfile change please.

Julesssss reviewed (changes requested):
Unfortunately there seems to have been a slight regression with the latest changes. As the user who creates a group chat, sends a message, and then an IOU split for that group, we should see the following Onyx data created for each individual 1:1
```
report_940
reportActions_940
reportIOUs_941       --> This is being created as reportIOUs_940
```

![Screenshot 2021-04-21 at 16 49 40](https://user-images.githubusercontent.com/10736861/115583449-cc9a8a80-a2c1-11eb-9b36-662f2792f13e.png)

So we're creating the chat reports, but not creating or linking to the IOUReport. Other than this and the issue where IOUReports are not retrievable via the push data, everything else is testing well.

Julesssss commented:
To round up the feedback:
- The only remaining bug to fix is that the I[OUReportID is being set to the ChatReportID](https://github.com/Expensify/Expensify.cash/pull/2112#pullrequestreview-641273841)
- The messaging[ here](https://github.com/Expensify/Expensify.cash/pull/2112#discussion_r617139091) should be updated, and the conditional check modified
- `Pod install` should be run in root/iOS and the change committed

Once that's done I'll ping the remaining reviewers and run the test flows once more.

tugbadogan commented:
> Unfortunately there seems to have been a slight regression with the latest changes. As the user who creates a group chat, sends a message, and then an IOU split for that group, we should see the following Onyx data created for each individual 1:1
> 
> So we're creating the chat reports, but not creating or linking to the IOUReport. Other than this and the issue where IOUReports are not retrievable via the push data, everything else is testing well.

I tested split bill flow for a group chat. The problem is caused by push notification being sent with `iouReportID` for each report in the group instead of `chatReportID`. You solved this issue with 1:1 IOU report creation.

> ```
> report_940
> reportActions_940
> reportIOUs_941       --> This is being created as reportIOUs_940
> ```

> ![Screenshot 2021-04-21 at 16 49 40](https://user-images.githubusercontent.com/10736861/115583449-cc9a8a80-a2c1-11eb-9b36-662f2792f13e.png)

In this example `940` is `iouReportID` so `reportIOUs_940` is created as expected. `report_940` and `reportActions_940` shouldn't be created with this id. If you can fix the push notifications being sent with `chatReportIDs`, this problem shouldn't occur.

tugbadogan commented:
@Julesssss 
This is request for IOU split action. I think it returns created IOU report IDs in `reportIDList`.

<img width="1440" alt="Screenshot 2021-04-22 at 00 45 51" src="https://user-images.githubusercontent.com/3853003/115636097-0a6cd280-a305-11eb-844e-245b6d00f744.png">

And we are getting push notifications for these IOU report IDs.

<img width="1440" alt="Screenshot 2021-04-22 at 00 51 37" src="https://user-images.githubusercontent.com/3853003/115636106-0e005980-a305-11eb-828a-4bfb5197f8e3.png">

When we call `Get` API method for `reportStuff` we get details about IOU report as expected.

<img width="1440" alt="Screenshot 2021-04-22 at 00 47 24" src="https://user-images.githubusercontent.com/3853003/115636100-0b9dff80-a305-11eb-948c-530b39d5c693.png">

tugbadogan committed:
Updated comment and pod install

Julesssss commented:
> I tested split bill flow for a group chat. The problem is caused by push notification being sent with `iouReportID` for each report in the group instead of `chatReportID`. You solved this issue with 1:1 IOU report creation.

Great spot! Thanks for pointing this out... I've just fixed this and will report back once this change has gone live on prod.

Julesssss commented on ios/ExpensifyCash.xcodeproj/project.pbxproj:
```diff
@@ -371,6 +373,24 @@
 			shellScript = "diff \"${PODS_PODFILE_DIR_PATH}/Podfile.lock\" \"${PODS_ROOT}/Manifest.lock\" > /dev/null\nif [ $? != 0 ] ; then\n    # print error to STDERR\n    echo \"error: The sandbox is not in sync with the Podfile.lock. Run 'pod install' or update your CocoaPods installation.\" >&2\n    exit 1\nfi\n# This output is used by Xcode 'outputs' to avoid re-running this script phase.\necho \"SU ...
 			showEnvVarsInLog = 0;
 		};
+		B6C8E0BF2CC222D8992DC951 /* [CP] Embed Pods Frameworks */ = {
+			isa = PBXShellScriptBuildPhase;
+			buildActionMask = 2147483647;
+			files = (
+			);
+			inputPaths = (
+				"${PODS_ROOT}/Target Support Files/Pods-ExpensifyCash-ExpensifyCashTests/Pods-ExpensifyCash-ExpensifyCashTests-frameworks.sh",
+				"${PODS_XCFRAMEWORKS_BUILD_DIR}/OpenSSL/OpenSSL.framework/OpenSSL",
+			);
+			name = "[CP] Embed Pods Frameworks";
+			outputPaths = (
+				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/OpenSSL.framework",
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "\"${PODS_ROOT}/Target Support Files/Pods-ExpensifyCash-ExpensifyCashTests/Pods-ExpensifyCash-ExpensifyCashTests-frameworks.sh\"\n";
+			showEnvVarsInLog = 0;
+		};
```
I don't think this change is required. I can't build the iOS app since this commit and I think it's because of these changes?

tugbadogan replied:
That's right, this is changed when I ran `pod install`, reverted these changes now.

tugbadogan committed:
Revert changes from pod install

tugbadogan committed:
Merge pull request #18 from Expensify/main

Merge

tugbadogan committed:
Merge branch 'master' into tugbadogan-create-iou-preview-component

Julesssss reviewed (dismissed):
Looking good. Not sure what happened with the last commit, but thanks for reverting. 

I've tested the flows against this internal [Web-E PR](https://github.com/Expensify/Expensify/issues/161333) which fixes the push notification problem we introduced.  It also requires pulling the latest Auth changes and rebuilding. 

I'll update the test steps to let QA know what is testable at this stage. I think for now this should be tested internally.

Julesssss commented:
@NikkiWines @marcaaron @AndrewGable -- not much has changed since your previous reviews, but I'll leave the PR open for a couple of hours in case you have any last thoughts.

iwiznia commented:
Seems you've got conflicts @tugbadogan, can you resolve them please?

tugbadogan committed:
Merge pull request #19 from Expensify/main

Merge

tugbadogan committed:
Merge branch 'master' into tugbadogan-create-iou-preview-component

tugbadogan commented:
Resolved the conflicts.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/Expensify.cash/actions/runs/822089210) to production in version: 1.0.39-5🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
